pkg/controller/release/strategy_executor.go

```diff
@@ -0,0 +1,470 @@
+			State:      newReleaseStrategyState,
+		}
+
+		previouslyAchievedStep := activeRelease.Status.AchievedStep
+		if previouslyAchievedStep == nil || targetStep != previouslyAchievedStep.Step {
+			// we validate that it fits in the len() of Strategy.Steps early in the process
+			targetStepName := activeRelease.Spec.Environment.Strategy.Steps[targetStep].Name
+			relStatus.AchievedStep = &shipper.AchievedStep{
+				Step: targetStep,
+				Name: targetStepName,
+			}
+			e.event(activeRelease, "step %d finished", targetStep)
+		}
+
+		if isLastStep {
+			condition := releaseutil.NewReleaseCondition(shipper.ReleaseConditionTypeComplete, corev1.ConditionTrue, "", "")
+			if diff := releaseutil.SetReleaseCondition(relStatus, *condition); !diff.IsEmpty() {
+				e.recorder.Eventf(
+					activeRelease,
+					corev1.EventTypeNormal,
+					"ReleaseConditionChanged",
+					diff.String())
+			}
+		}
+
+		if !equality.Semantic.DeepEqual(activeRelease.Status, *relStatus) {
+			releasePatches = append(releasePatches, &ReleaseUpdateResult{
+				NewStatus: relStatus,
+				Name:      activeRelease.Name,
+			})
+		}
+
+		return PipelineBreak, releasePatches, releaseStrategyStateTransitions, nil
+	}
+}
+
+/*
+	For each release object:
+	0. Ensure release scheduled.
+	  0.1. Choose clusters.
+	  0.2. Ensure target objects exist.
+	    0.2.1. Compare chosen clusters and if different, update the spec.
+	1. Find it's ancestor.
+	2. For the head release, ensure installation.
+	  2.1. Simply check installation targets.
+	3. For the head release, ensure capacity.
+	  3.1. Ensure the capacity corresponds to the strategy contender.
+	4. For the head release, ensure traffic.
+	  4.1. Ensure the traffic corresponds to the strategy contender.
+	5. For a tail release, ensure traffic.
+	  5.1. Look at the leader and check it's target traffic.
+	  5.2. Look at the strategy and figure out the target traffic.
+	6. For a tail release, ensure capacity.
+	  6.1. Look at the leader and check it's target capacity.
+	  6.2 Look at the strategy and figure out the target capacity.
+	7. Make necessary adjustments to the release object.
+*/
+
+func (e *StrategyExecutor) Execute() ([]ExecutorResult, []ReleaseStrategyStateTransition, error) {
+	var releaseStrategyConditions []shipper.ReleaseStrategyCondition
+	if e.curr.release.Status.Strategy != nil {
+		releaseStrategyConditions = e.curr.release.Status.Strategy.Conditions
+	}
+	cond := conditions.NewStrategyConditions(releaseStrategyConditions...)
+
+	isHead, hasTail := e.succ == nil, e.prev != nil
+
+	pipeline := NewPipeline()
+	if isHead {
+		pipeline.Enqueue(genInstallationEnforcer(e.curr, nil))
+	}
+	pipeline.Enqueue(genCapacityEnforcer(e.curr, e.succ))
+	pipeline.Enqueue(genTrafficEnforcer(e.curr, e.succ))
+
+	if isHead {
+		if hasTail {
+			pipeline.Enqueue(genTrafficEnforcer(e.prev, e.curr))
+			pipeline.Enqueue(genCapacityEnforcer(e.prev, e.curr))
+		}
+		pipeline.Enqueue(genReleaseStrategyStateEnforcer(e.curr, nil))
+	}
```

Comment:
I have nothing to add, but I just love this ❤️Both the pipeline itself and the enforcer generators. This solves a major gripe I had with the previous executor, which made refactoring each step a pain in the ass.